# Importing our dataframe

In [19]:
import pandas as pd

In [20]:
houses = pd.read_csv("kc_house_data.csv")
houses

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [21]:
type(houses) #Our file is a dataframe

pandas.core.frame.DataFrame

# Analizing the comun labes of our Dataframe

In [45]:
houses.columns 

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

One of the issues that we first noticed is that between the columns "view" and "grade", there is a column that it is not being shown.
That column is the **condition** column.

# Analizing the number of rows of the dataframe

In [46]:
len(houses)

21613

In [47]:
houses.shape

(21613, 21)

Our dataframe contains 21613 rows and 21 columns.

In [48]:
houses.size

453873

453873 is the area of our dataframe **Area=(rows*columns).**

# Asking Pandas to show us at least 20 rows

In [49]:
pd.options.display.min_rows = 20

In [50]:
houses

id             date      price  bedrooms  bathrooms  \
0      7129300520  20141013T000000   221900.0         3       1.00   
1      6414100192  20141209T000000   538000.0         3       2.25   
2      5631500400  20150225T000000   180000.0         2       1.00   
3      2487200875  20141209T000000   604000.0         4       3.00   
4      1954400510  20150218T000000   510000.0         3       2.00   
5      7237550310  20140512T000000  1225000.0         4       4.50   
6      1321400060  20140627T000000   257500.0         3       2.25   
7      2008000270  20150115T000000   291850.0         3       1.50   
8      2414600126  20150415T000000   229500.0         3       1.00   
9      3793500160  20150312T000000   323000.0         3       2.50   
...           ...              ...        ...       ...        ...   
21603  7852140040  20140825T000000   507250.0         3       2.50   
21604  9834201367  20150126T000000   429000.0         3       2.00   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1007500.0         4       3.50   
21607  2997800021  20150219T000000   475000.0         3       2.50   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
5             5420    101930     1.0           0     0  ...     11   
6             1715      6819     2.0           0     0  ...      7   
7             1060      9711     1.0           0     0  ...      7   
8             1780      7470     1.0           0     0  ...      7   
9             1890      6560     2.0           0     0  ...      7   
...            ...       ...     ...         ...   ...  ...    ...   
21603         2270      5536     2.0           0     0  ...      8   
21604         1490      1126     3.0           0     0  ...      8   
21605         2520      6023     2.0           0     0  ...      9   
21606         3510      7200     2.0           0     0  ...      9   
21607         1310      1294     2.0           0     0  ...      8   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
5            3890           1530      2001             0    98053  47.6561   
6            1715              0      1995             0    98003  47.3097   
7            1060              0      1963             0    98198  47.4095   
8            1050            730      1960             0    98146  47.5123   
9            1890              0      2003             0    98038  47.3684   
...           ...            ...       ...           ...      ...      ... 

# Creating a new dataframe that allows us to analyze the head of the original dataset

In [51]:
houses.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

# Datatypes & info()

In [52]:
houses.info() #General info of our dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [53]:
houses.dtypes #ONLY shows the datatypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

# Analizing the last 15 rows of the dataframe

In [54]:
houses.tail(15)

id             date      price  bedrooms  bathrooms  \
21598  8956200760  20141013T000000   541800.0         4       2.50   
21599  7202300110  20140915T000000   810000.0         4       3.00   
21600   249000205  20141015T000000  1537000.0         5       3.75   
21601  5100403806  20150407T000000   467000.0         3       2.50   
21602   844000965  20140626T000000   224000.0         3       1.75   
21603  7852140040  20140825T000000   507250.0         3       2.50   
21604  9834201367  20150126T000000   429000.0         3       2.00   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1007500.0         4       3.50   
21607  2997800021  20150219T000000   475000.0         3       2.50   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21598         3118      7866     2.0           0     2  ...      9   
21599         3990      7838     2.0           0     0  ...      9   
21600         4470      8088     2.0           0     0  ...     11   
21601         1425      1179     3.0           0     0  ...      8   
21602         1500     11968     1.0           0     0  ...      6   
21603         2270      5536     2.0           0     0  ...      8   
21604         1490      1126     3.0           0     0  ...      8   
21605         2520      6023     2.0           0     0  ...      9   
21606         3510      7200     2.0           0     0  ...      9   
21607         1310      1294     2.0           0     0  ...      8   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21598        3118              0      2014             0    98001  47.2931   
21599        3990              0      2003             0    98053  47.6857   
21600        4470              0      2008             0    98004  47.6321   
21601        1425              0      2008             0    98125  47.6963   
21602        1500              0      2014             0    98010  47.3095   
21603        2270              0      2003             0    98065  47.5389   
21604        1490              0      2014             0    98144  47.5699   
21605        2520              0      2014             0    98056  47.5137   
21606        2600            910      2009             0    98136  47.5537   
21607        1180            130      2008             0    98116  47.5773   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21598 -122.264           2673        6500  
21599 -122.046           3370        6814  
21600 -122.200           2780        8964  
21601 -122.318           1285        1253  
21602 -122.002           1320       11303  
21603 -121.881           2270        5731  
21604 -122.288           1400        1230  
21605 -122.167           2520        6023  
21606 -122.398           2050        6200  
21607 -122.409           1330        1265  
21608 -122.346           1530        150

# Analizing the first 5 rows of the dataset

In [55]:
houses.head(5)

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

# Basic DataFrame Computations

+ Sum
+ Min
+ Max
+ Median
+ Mode
+ Mean 
+ Count
+ Describe

In [56]:
houses

id             date      price  bedrooms  bathrooms  \
0      7129300520  20141013T000000   221900.0         3       1.00   
1      6414100192  20141209T000000   538000.0         3       2.25   
2      5631500400  20150225T000000   180000.0         2       1.00   
3      2487200875  20141209T000000   604000.0         4       3.00   
4      1954400510  20150218T000000   510000.0         3       2.00   
5      7237550310  20140512T000000  1225000.0         4       4.50   
6      1321400060  20140627T000000   257500.0         3       2.25   
7      2008000270  20150115T000000   291850.0         3       1.50   
8      2414600126  20150415T000000   229500.0         3       1.00   
9      3793500160  20150312T000000   323000.0         3       2.50   
...           ...              ...        ...       ...        ...   
21603  7852140040  20140825T000000   507250.0         3       2.50   
21604  9834201367  20150126T000000   429000.0         3       2.00   
21605  3448900210  20141014T000000   610685.0         4       2.50   
21606  7936000429  20150326T000000  1007500.0         4       3.50   
21607  2997800021  20150219T000000   475000.0         3       2.50   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
5             5420    101930     1.0           0     0  ...     11   
6             1715      6819     2.0           0     0  ...      7   
7             1060      9711     1.0           0     0  ...      7   
8             1780      7470     1.0           0     0  ...      7   
9             1890      6560     2.0           0     0  ...      7   
...            ...       ...     ...         ...   ...  ...    ...   
21603         2270      5536     2.0           0     0  ...      8   
21604         1490      1126     3.0           0     0  ...      8   
21605         2520      6023     2.0           0     0  ...      9   
21606         3510      7200     2.0           0     0  ...      9   
21607         1310      1294     2.0           0     0  ...      8   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
5            3890           1530      2001             0    98053  47.6561   
6            1715              0      1995             0    98003  47.3097   
7            1060              0      1963             0    98198  47.4095   
8            1050            730      1960             0    98146  47.5123   
9            1890              0      2003             0    98038  47.3684   
...           ...            ...       ...           ...      ...      ... 

## Min & Max

In [57]:
houses.min()  #Output is the column + Min. Value

id                       1000102
date             20140502T000000
price                    75000.0
bedrooms                       0
bathrooms                    0.0
sqft_living                  290
sqft_lot                     520
floors                       1.0
waterfront                     0
view                           0
condition                      1
grade                          1
sqft_above                   290
sqft_basement                  0
yr_built                    1900
yr_renovated                   0
zipcode                    98001
lat                      47.1559
long                    -122.519
sqft_living15                399
sqft_lot15                   651
dtype: object

In [58]:
houses.max()    #Output is the column + Max. Value

id                    9900000190
date             20150527T000000
price                  7700000.0
bedrooms                      33
bathrooms                    8.0
sqft_living                13540
sqft_lot                 1651359
floors                       3.5
waterfront                     1
view                           4
condition                      5
grade                         13
sqft_above                  9410
sqft_basement               4820
yr_built                    2015
yr_renovated                2015
zipcode                    98199
lat                      47.7776
long                    -121.315
sqft_living15               6210
sqft_lot15                871200
dtype: object

In [59]:
type(houses.max())

pandas.core.series.Series

**Remember** Series is a Pandas´s data structure

# Summing all numeric values

In [60]:
houses.sum(numeric_only=True)

id               9.899406e+13
price            1.167293e+10
bedrooms         7.285400e+04
bathrooms        4.570625e+04
sqft_living      4.495287e+07
sqft_lot         3.265069e+08
floors           3.229650e+04
waterfront       1.630000e+02
view             5.064000e+03
condition        7.368800e+04
grade            1.654880e+05
sqft_above       3.865249e+07
sqft_basement    6.300385e+06
yr_built         4.259933e+07
yr_renovated     1.824186e+06
zipcode          2.119759e+09
lat              1.027915e+06
long            -2.641409e+06
sqft_living15    4.293536e+07
sqft_lot15       2.759646e+08
dtype: float64

In [61]:
2.759646e+08  #changing the notation

275964600.0

# Getting the number of values present per column

In [62]:
houses.count()

id               21613
date             21613
price            21613
bedrooms         21613
bathrooms        21613
sqft_living      21613
sqft_lot         21613
floors           21613
waterfront       21613
view             21613
condition        21613
grade            21613
sqft_above       21613
sqft_basement    21613
yr_built         21613
yr_renovated     21613
zipcode          21613
lat              21613
long             21613
sqft_living15    21613
sqft_lot15       21613
dtype: int64

# Mean, Median, & Mode

## Mean

In [74]:
houses.mean(numeric_only=True)

id               4.580302e+09
price            5.400881e+05
bedrooms         3.370842e+00
bathrooms        2.114757e+00
sqft_living      2.079900e+03
sqft_lot         1.510697e+04
floors           1.494309e+00
waterfront       7.541757e-03
view             2.343034e-01
condition        3.409430e+00
grade            7.656873e+00
sqft_above       1.788391e+03
sqft_basement    2.915090e+02
yr_built         1.971005e+03
yr_renovated     8.440226e+01
zipcode          9.807794e+04
lat              4.756005e+01
long            -1.222139e+02
sqft_living15    1.986552e+03
sqft_lot15       1.276846e+04
dtype: float64

In [75]:
 5.400881e+05


540088.1

The average price of a home in King County WA is **540088.1USD**

## Median

In [78]:
houses.median(numeric_only=True)

id               3.904930e+09
price            4.500000e+05
bedrooms         3.000000e+00
bathrooms        2.250000e+00
sqft_living      1.910000e+03
sqft_lot         7.618000e+03
floors           1.500000e+00
waterfront       0.000000e+00
view             0.000000e+00
condition        3.000000e+00
grade            7.000000e+00
sqft_above       1.560000e+03
sqft_basement    0.000000e+00
yr_built         1.975000e+03
yr_renovated     0.000000e+00
zipcode          9.806500e+04
lat              4.757180e+01
long            -1.222300e+02
sqft_living15    1.840000e+03
sqft_lot15       7.620000e+03
dtype: float64

In [79]:
4.500000e+05

450000.0

The middle price of a house in King county WA is **450000.0USD**

## Mode

In [88]:
houses.mode(numeric_only=True)

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  795000620.0  350000.0       3.0        2.5       1300.0    5000.0     1.0   
1          NaN  450000.0       NaN        NaN          NaN       NaN     NaN   
2          NaN       NaN       NaN        NaN          NaN       NaN     NaN   
3          NaN       NaN       NaN        NaN          NaN       NaN     NaN   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0         0.0   0.0        3.0    7.0      1300.0            0.0    2014.0   
1         NaN   NaN        NaN    NaN         NaN            NaN       NaN   
2         NaN   NaN        NaN    NaN         NaN            NaN       NaN   
3         NaN   NaN        NaN    NaN         NaN            NaN       NaN   

   yr_renovated  zipcode      lat    long  sqft_living15  sqft_lot15  
0           0.0  98103.0  47.5322 -122.29         1540.0      5000.0  
1           NaN      NaN  47.5491     NaN            NaN         NaN  
2           NaN      NaN  47.6624     NaN            NaN         NaN  
3           NaN      NaN  47.6846     NaN            NaN         NaN

# The Describe Method

The describe method is used to generate **descriptive statistics**.

In [92]:
houses.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [101]:
houses.describe(include=["O"])  

date
count             21613
unique              372
top     20140623T000000
freq                142

The line of code above does not make sense. However, it is a line that i can reference whenever I want to get descriptive stats from objects.

## Getting the descriptive stats of integer data types

In [103]:
houses.describe(include=["int64"])

id      bedrooms   sqft_living      sqft_lot    waterfront  \
count  2.161300e+04  21613.000000  21613.000000  2.161300e+04  21613.000000   
mean   4.580302e+09      3.370842   2079.899736  1.510697e+04      0.007542   
std    2.876566e+09      0.930062    918.440897  4.142051e+04      0.086517   
min    1.000102e+06      0.000000    290.000000  5.200000e+02      0.000000   
25%    2.123049e+09      3.000000   1427.000000  5.040000e+03      0.000000   
50%    3.904930e+09      3.000000   1910.000000  7.618000e+03      0.000000   
75%    7.308900e+09      4.000000   2550.000000  1.068800e+04      0.000000   
max    9.900000e+09     33.000000  13540.000000  1.651359e+06      1.000000   

               view     condition         grade    sqft_above  sqft_basement  \
count  21613.000000  21613.000000  21613.000000  21613.000000   21613.000000   
mean       0.234303      3.409430      7.656873   1788.390691     291.509045   
std        0.766318      0.650743      1.175459    828.090978     442.575043   
min        0.000000      1.000000      1.000000    290.000000       0.000000   
25%        0.000000      3.000000      7.000000   1190.000000       0.000000   
50%        0.000000      3.000000      7.000000   1560.000000       0.000000   
75%        0.000000      4.000000      8.000000   2210.000000     560.000000   
max        4.000000      5.000000     13.000000   9410.000000    4820.000000   

           yr_built  yr_renovated       zipcode  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean    1971.005136     84.402258  98077.939805    1986.552492   12768.455652  
std       29.373411    401.679240     53.505026     685.391304   27304.179631  
min     1900.000000      0.000000  98001.000000     399.000000     651.000000  
25%     1951.000000      0.000000  98033.000000    1490.000000    5100.000000  
50%     1975.000000      0.000000  98065.000000    1840.000000    7620.000000  
75%     1997.000000      0.000000  98118.000000    2360.000000   10083.000000  
max     2015.000000   2015.000000  98199.000000    6210.000000  871200.000000

# Series & Columns

Instead of working with an entire dataframe - we will go ahead and work with one piece (column)

**series** is a pandas data structure.

## First way of slecting a column

**df.age**.  <-- Not very reliable

## Second way of doing so

**df["age"]** <-- very reliable

# Implementing what we learned using the Titanic and Netflix datasets

In [27]:
import pandas as pd
titanic = pd.read_csv("titanic.csv")
netflix = pd.read_csv("netflix_titles.csv", sep="|", index_col=0)
houses = pd.read_csv("kc_house_data.csv")

In [9]:
titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

      age  sibsp  parch  ticket      fare    cabin embarked boat body  \
0      29      0      0   24160  211.3375       B5        S    2    ?   
1  0.9167      1      2  113781    151.55  C22 C26        S   11    ?   
2       2      1      2  113781    151.55  C22 C26        S    ?    ?   
3      30      1      2  113781    151.55  C22 C26        S    ?  135   
4      25      1      2  113781    151.55  C22 C26        S    ?    ?   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

Lets suppose that we want to work just with the **name** column.

In [11]:
titanic.name

0                         Allen, Miss. Elisabeth Walton
1                        Allison, Master. Hudson Trevor
2                          Allison, Miss. Helen Loraine
3                  Allison, Mr. Hudson Joshua Creighton
4       Allison, Mrs. Hudson J C (Bessie Waldo Daniels)
                             ...                       
1304                               Zabour, Miss. Hileni
1305                              Zabour, Miss. Thamine
1306                          Zakarian, Mr. Mapriededer
1307                                Zakarian, Mr. Ortin
1308                                 Zimmerman, Mr. Leo
Name: name, Length: 1309, dtype: object

## Figuring out the data type of the data structure

In [12]:
type(titanic.name)

pandas.core.series.Series

Our data type is series with a length of 1309. Which means that the dataset has 1309 rown.

In [13]:
titanic["name"]

0                         Allen, Miss. Elisabeth Walton
1                        Allison, Master. Hudson Trevor
2                          Allison, Miss. Helen Loraine
3                  Allison, Mr. Hudson Joshua Creighton
4       Allison, Mrs. Hudson J C (Bessie Waldo Daniels)
                             ...                       
1304                               Zabour, Miss. Hileni
1305                              Zabour, Miss. Thamine
1306                          Zakarian, Mr. Mapriededer
1307                                Zakarian, Mr. Ortin
1308                                 Zimmerman, Mr. Leo
Name: name, Length: 1309, dtype: object

In [14]:
titanic.age

0           29
1       0.9167
2            2
3           30
4           25
         ...  
1304      14.5
1305         ?
1306      26.5
1307        27
1308        29
Name: age, Length: 1309, dtype: object

In [15]:
titanic["age"]

0           29
1       0.9167
2            2
3           30
4           25
         ...  
1304      14.5
1305         ?
1306      26.5
1307        27
1308        29
Name: age, Length: 1309, dtype: object

# A closer look at series

A **series** is a one-dimensional ndarray with axis labels. Their numeric index will start at number 0.

A **dataframe** is a 2D data structure

In [40]:
houses.sum(numeric_only=True)  #We are summing the values for a given column

id               9.899406e+13
price            1.167293e+10
bedrooms         7.285400e+04
bathrooms        4.570625e+04
sqft_living      4.495287e+07
sqft_lot         3.265069e+08
floors           3.229650e+04
waterfront       1.630000e+02
view             5.064000e+03
condition        7.368800e+04
grade            1.654880e+05
sqft_above       3.865249e+07
sqft_basement    6.300385e+06
yr_built         4.259933e+07
yr_renovated     1.824186e+06
zipcode          2.119759e+09
lat              1.027915e+06
long            -2.641409e+06
sqft_living15    4.293536e+07
sqft_lot15       2.759646e+08
dtype: float64

# Let´s get the total price of all the houses sold in king county in 2014 - 2015

In [45]:
houses.price.sum()

11672925008.0

# Let´s get the minimun price

In [46]:
houses.price.min()

75000.0

Series have different properties

**For example**, the **shape** property shows us the number of rown and the number of columns in 1D.

# Let´s look at the index

In [49]:
houses.index

RangeIndex(start=0, stop=21613, step=1)

In [54]:
mins = houses.min(numeric_only=True)

In [56]:
mins.values #Here is an array with all the values of the cheapest home´s price in king county.

array([ 1.000102e+06,  7.500000e+04,  0.000000e+00,  0.000000e+00,
        2.900000e+02,  5.200000e+02,  1.000000e+00,  0.000000e+00,
        0.000000e+00,  1.000000e+00,  1.000000e+00,  2.900000e+02,
        0.000000e+00,  1.900000e+03,  0.000000e+00,  9.800100e+04,
        4.715590e+01, -1.225190e+02,  3.990000e+02,  6.510000e+02])

In [58]:
mins.index   #We get a 1D data with labels

Index(['id', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')

# Important series methods

+ head()
+ tail()
+ describe()
+ unique()
+ nunique()
+ nlargest()
+ nsmallest()
+ value_counts()
+ plot()

# unique & nunique

They are all series methods.

In [59]:
houses["bedrooms"] #This will give us the number of bedrooms in each house.


0        3
1        3
2        2
3        4
4        3
        ..
21608    3
21609    4
21610    2
21611    3
21612    2
Name: bedrooms, Length: 21613, dtype: int64

Let´s go ahead and get the **unique** number of values.

In [63]:
houses["bedrooms"].unique()

array([ 3,  2,  4,  5,  1,  6,  7,  0,  8,  9, 11, 10, 33])

# Now, we want to know a number of how many unique bedrooms are in a series

In [68]:
houses.bedrooms.nunique()
# We have 13 unique values in this dataset

13

# Now, let´s figure out how many unique zip codes are there in King County WA

In [71]:
houses.zipcode.nunique()
# there are 70 unique zip codes in this King County´s dataset

70

# The **nunique** method does have a parameter known as **dropna**

The netflic dataset does have missing values.

In [73]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 894.5+ KB


In [76]:
netflix.rating.nunique() #The missing values are not being included in here.

17

In [79]:
netflix.rating.nunique(dropna=False) #This calculation includes missing values

18